File: ALittle/Module/ALittleIDE/Other/GameLibrary/SEngine/Script/Utility/ClientSystem.lua

module("ALittle", package.seeall)

local ___rawset = rawset
local ___pairs = pairs
local ___ipairs = ipairs
local ___all_struct = GetAllStruct()

RegStruct(1715346212, "ALittle.Event", {
name = "ALittle.Event", ns_name = "ALittle", rl_name = "Event", hash_code = 1715346212,
name_list = {"target"},
type_list = {"ALittle.EventDispatcher"},
option_map = {}
})
RegStruct(-1221484301, "ALittle.ClientConnectEvent", {
name = "ALittle.ClientConnectEvent", ns_name = "ALittle", rl_name = "ClientConnectEvent", hash_code = -1221484301,
name_list = {"target","msg_receiver"},
type_list = {"ALittle.EventDispatcher","ALittle.MsgReceiverTemplate<ALittle.MsgReceiverNative,lua.__CPPAPIMessageWriteFactory>"},
option_map = {}
})
RegStruct(-245025090, "ALittle.ClientDisconnectEvent", {
name = "ALittle.ClientDisconnectEvent", ns_name = "ALittle", rl_name = "ClientDisconnectEvent", hash_code = -245025090,
name_list = {"target","msg_receiver"},
type_list = {"ALittle.EventDispatcher","ALittle.MsgReceiverTemplate<ALittle.MsgReceiverNative,lua.__CPPAPIMessageWriteFactory>"},
option_map = {}
})

assert(ALittle.IMsgCommonNative, " extends class:ALittle.IMsgCommonNative is nil")
MsgReceiverNative = Lua.Class(ALittle.IMsgCommonNative, "ALittle.MsgReceiverNative")

function MsgReceiverNative:SetID(id)
	self._client_id = id
end

function MsgReceiverNative:GetID()
	return self._client_id
end

function MsgReceiverNative:SendFactory(factory)
	__CPPAPI_ServerSchedule:ClientSend(self._client_id, factory)
end

function MsgReceiverNative:Close()
	__CPPAPI_ServerSchedule:ClientClose(self._client_id)
	A_ClientSystem:RemoveMsgServer(self._client_id)
end

MsgReceiver = Lua.Template(MsgReceiverTemplate, "ALittle.MsgReceiverTemplate<ALittle.MsgReceiverNative, lua.__CPPAPIMessageWriteFactory>", MsgReceiverNative, __CPPAPIMessageWriteFactory);
assert(ALittle.EventDispatcher, " extends class:ALittle.EventDispatcher is nil")
ClientSystem = Lua.Class(ALittle.EventDispatcher, "ALittle.ClientSystem")

function ClientSystem:Ctor()
	___rawset(self, "_client_map", {})
end

function ClientSystem:RemoveMsgServer(client_id)
	self._client_map[client_id] = nil
end

function ClientSystem:HandleClientConnect(client_id, remote_ip, remote_port)
	local client = MsgReceiver(client_id, remote_ip, remote_port)
	self._client_map[client_id] = client
	client:HandleConnected()
	local event = {}
	event.target = self
	event.msg_receiver = client
	self:DispatchEvent(___all_struct[-1221484301], event)
end

function ClientSystem:HandleClientDisconnect(client_id)
	local client = self._client_map[client_id]
	if client == nil then
		return
	end
	self._client_map[client_id] = nil
	client:HandleDisconnected()
	local event = {}
	event.target = self
	event.msg_receiver = client
	self:DispatchEvent(___all_struct[-245025090], event)
end

function ClientSystem:HandleClientMessage(client_id, id, rpc_id, factory)
	local client = self._client_map[client_id]
	if client == nil then
		return
	end
	client:HandleMessage(id, rpc_id, factory)
end

_G.A_ClientSystem = ClientSystem()

File: ALittle/Module/ALittleIDE/Other/GameLibrary/SEngine/JSScript/Utility/ClientSystem.js
{
if (typeof ALittle === "undefined") ALittle = {};
let ___all_struct = ALittle.GetAllStruct();

ALittle.RegStruct(1715346212, "ALittle.Event", {
name : "ALittle.Event", ns_name : "ALittle", rl_name : "Event", hash_code : 1715346212,
name_list : ["target"],
type_list : ["ALittle.EventDispatcher"],
option_map : {}
})
ALittle.RegStruct(-1221484301, "ALittle.ClientConnectEvent", {
name : "ALittle.ClientConnectEvent", ns_name : "ALittle", rl_name : "ClientConnectEvent", hash_code : -1221484301,
name_list : ["target","msg_receiver"],
type_list : ["ALittle.EventDispatcher","ALittle.MsgReceiverTemplate<ALittle.MsgReceiverNative,lua.__CPPAPIMessageWriteFactory>"],
option_map : {}
})
ALittle.RegStruct(-245025090, "ALittle.ClientDisconnectEvent", {
name : "ALittle.ClientDisconnectEvent", ns_name : "ALittle", rl_name : "ClientDisconnectEvent", hash_code : -245025090,
name_list : ["target","msg_receiver"],
type_list : ["ALittle.EventDispatcher","ALittle.MsgReceiverTemplate<ALittle.MsgReceiverNative,lua.__CPPAPIMessageWriteFactory>"],
option_map : {}
})

if (ALittle.IMsgCommonNative === undefined) throw new Error(" extends class:ALittle.IMsgCommonNative is undefined");
ALittle.MsgReceiverNative = JavaScript.Class(ALittle.IMsgCommonNative, {
	SetID : function(id) {
		this._client_id = id;
	},
	GetID : function() {
		return this._client_id;
	},
	SendFactory : function(factory) {
		__CPPAPI_ServerSchedule.ClientSend(this._client_id, factory);
	},
	Close : function() {
		__CPPAPI_ServerSchedule.ClientClose(this._client_id);
		A_ClientSystem.RemoveMsgServer(this._client_id);
	},
}, "ALittle.MsgReceiverNative");

MsgReceiver = JavaScript.Template(ALittle.MsgReceiverTemplate, "ALittle.MsgReceiverTemplate<ALittle.MsgReceiverNative, lua.__CPPAPIMessageWriteFactory>", ALittle.MsgReceiverNative, lua.__CPPAPIMessageWriteFactory);
if (ALittle.EventDispatcher === undefined) throw new Error(" extends class:ALittle.EventDispatcher is undefined");
ALittle.ClientSystem = JavaScript.Class(ALittle.EventDispatcher, {
	Ctor : function() {
		this._client_map = new Map();
	},
	RemoveMsgServer : function(client_id) {
		this._client_map.delete(client_id);
	},
	HandleClientConnect : function(client_id, remote_ip, remote_port) {
		let client = ALittle.NewObject(ALittle.MsgReceiver, client_id, remote_ip, remote_port);
		this._client_map.set(client_id, client);
		client.HandleConnected();
		let event = {};
		event.target = this;
		event.msg_receiver = client;
		this.DispatchEvent(___all_struct.get(-1221484301), event);
	},
	HandleClientDisconnect : function(client_id) {
		let client = this._client_map.get(client_id);
		if (client === undefined) {
			return;
		}
		this._client_map.delete(client_id);
		client.HandleDisconnected();
		let event = {};
		event.target = this;
		event.msg_receiver = client;
		this.DispatchEvent(___all_struct.get(-245025090), event);
	},
	HandleClientMessage : function(client_id, id, rpc_id, factory) {
		let client = this._client_map.get(client_id);
		if (client === undefined) {
			return;
		}
		client.HandleMessage(id, rpc_id, factory);
	},
}, "ALittle.ClientSystem");

window.A_ClientSystem = ALittle.NewObject(ALittle.ClientSystem);
}